src/network/networkd-routing-policy-rule.c

```
@@ -0,0 +1,859 @@
+#include "socket-util.h"
+#include "string-util.h"
+
+int routing_policy_rule_new(RoutingPolicyRule **ret) {
+        RoutingPolicyRule *rule = NULL;
+
+        rule = new0(RoutingPolicyRule, 1);
+        if (!rule)
+                return -ENOMEM;
+
+        rule->family = AF_INET;
+        rule->table = RT_TABLE_MAIN;
+
+        *ret = rule;
+        rule = NULL;
+
+        return 0;
+}
+
+void routing_policy_rule_free(RoutingPolicyRule *rule) {
+
+        if (!rule)
+                return;
+
+        if (rule->network) {
+                LIST_REMOVE(rules, rule->network->rules, rule);
+                assert(rule->network->n_rules > 0);
+                rule->network->n_rules--;
+
+                if (rule->section) {
+                        hashmap_remove(rule->network->rules_by_section, rule->section);
+                        network_config_section_free(rule->section);
+                }
+
+                if (rule->network->manager) {
+                        set_remove(rule->network->manager->rules, rule);
+                        set_remove(rule->network->manager->rules_foreign, rule);
+                }
+        }
+
+        free(rule);
+}
+
+static void routing_policy_rule_hash_func(const void *b, struct siphash *state) {
+        const RoutingPolicyRule *rule = b;
+
+        assert(rule);
+
+        siphash24_compress(&rule->family, sizeof(rule->family), state);
+
+        switch (rule->family) {
+        case AF_INET:
+        case AF_INET6:
+
+                if (!in_addr_is_null(rule->family, &rule->from)) {
+                        siphash24_compress(&rule->from, FAMILY_ADDRESS_SIZE(rule->family), state);
+                        siphash24_compress(&rule->from_prefixlen, sizeof(rule->from_prefixlen), state);
+                } else {
+                        siphash24_compress(&rule->to, FAMILY_ADDRESS_SIZE(rule->family), state);
+                        siphash24_compress(&rule->to_prefixlen, sizeof(rule->to_prefixlen), state);
+                }
```

Hm, if `from` is present, `to` is ignored and rules with different `to` hash the same. Is this on purpose?

And similarly in routing_policy_rule_compare_func below...

---

yes Dont think from and to is used together. Could not find any example.

---

`ip` certainly support this. Your code also does not disallow it, just silently skips parts of the rule. Please just make it support both or one or none in the same rule.

---

Please give me a example . need to test it

---

`ip rule add from 192.168.120.18/24 to 192.168.121.18/24 lookup 123`